src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -1284,7 +1254,70 @@ private void SetFlagToAddListSeparatorBeforeNextItem()
+            => ThrowHelper.ThrowInvalidOperationException(ExceptionResource.DepthTooLarge, _currentDepth, _options.MaxDepth, token: default, tokenType: default);
+
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
         private string DebuggerDisplay => $"BytesCommitted = {BytesCommitted} BytesPending = {BytesPending} CurrentDepth = {CurrentDepth}";
+
+        /// <summary>
+        /// Indicates whether the writer is currently writing a partial string value.
+        /// </summary>
+        private bool IsWritingPartialString => _enclosingContainer >= EnclosingContainerType.Utf8StringSequence;
+
+        /// <summary>
+        /// The type of container that is enclosing the current position. The underlying values have been chosen
+        /// to allow <see cref="CanWriteValue"/> to be done using bitwise operations and must be kept in sync with <see cref="JsonTokenType"/>.
+        /// </summary>
+        internal enum EnclosingContainerType : byte
+        {
+            /// <summary>
+            /// Root level.
+            /// </summary>
+            None = JsonTokenType.None,
+
+            /// <summary>
+            /// JSON object.
+            /// </summary>
+            Object = JsonTokenType.PropertyName,
+
+            /// <summary>
+            /// JSON array. This is chosen large enough to not conflict with any <see cref="JsonTokenType"/> value.
+            /// </summary>
+            Array =                 0b000_1_0000,
+
+            /// <summary>
+            /// Partial UTF-8 string. This is a container if viewed as an array of "utf-8 string segment"-typed values. This array can only be one level deep
+            /// so <see cref="_bitStack"/> does not need to store its state.
+            /// <see cref="IsWritingPartialString"/> relies on the value of the partial string members being the largest values of this enum.
+            /// </summary>
+            Utf8StringSequence =    0b001_0_0000,
+
+            /// <summary>
+            /// Partial UTF-16 string. This is a container if viewed as an array of "utf-16 string segment"-typed values. This array can only be one level deep
+            /// so <see cref="_bitStack"/> does not need to store its state.
+            /// <see cref="IsWritingPartialString"/> relies on the value of the partial string members being the largest values of this enum.
+            /// </summary>
+            Utf16StringSequence =   0b010_0_0000,
+
+            /// <summary>
+            /// Partial Base64 string. This is a container if viewed as an array of "base64 string segment"-typed values. This array can only be one level deep
+            /// so <see cref="_bitStack"/> does not need to store its state.
+            /// <see cref="IsWritingPartialString"/> relies on the value of the partial string members being the largest values of this enum.
+            /// </summary>
+            Base64StringSequence =  0b011_0_0000,
```

Comment:
Remind me, why do the bits of this value overlap with the previous two container types? Couldn't it just be an independent value like `1 << 7`?

Reply:
That made it seem like no more values could be added to the enum, but actually an equivalently simple encoding that would still work is 1<<4, 2<<4, 3<<4 and 4<<4 for these last 4 members.

Reply:
Presumably you mean `1 <<4`, `1 << 5`, `1 << 6`, etc?

Reply:
No, I meant they don't need to be flags-style as long as they're >= 16 (avoiding bit overlap with json token type).

Reply:
Changed to hex for readability and renumbered to 0x10, 0x20, 0x30 and 0x40.